### PR TITLE
Spring 7 hw 41

### PR DESCRIPTION
**Pull request by yurwing**

### CODE REVIEW

File: src/main/java/com/dev/cinema/controller/UserController.java [hunk 1]
```diff
@@ -22,7 +25,7 @@ public UserController(UserService userService, UserMapper userMapper) {
     }
 
     @GetMapping("/by-email")
-    public UserResponseDto get(@RequestParam String email) {
+    public UserResponseDto get(@RequestParam @NotNull @Size(min = 4) @Valid String email) {
```
Comment (Max-Pochepets) [inline src/main/java/com/dev/cinema/controller/UserController.java]: As I know, `@NotNull` is redundant here, cause `@RequestParam` without `required` will throw exception if you send null as parameter. @Size is OK

File: src/main/java/com/dev/cinema/model/dto/request/CinemaHallRequestDto.java [hunk 2]
```diff
@@ -1,7 +1,12 @@
 package com.dev.cinema.model.dto.request;
 
+import javax.validation.constraints.NotEmpty;
+import javax.validation.constraints.NotNull;
+
 public class CinemaHallRequestDto {
+    @NotEmpty
```
Comment (Max-Pochepets) [inline src/main/java/com/dev/cinema/model/dto/request/CinemaHallRequestDto.java]: [javadoc](https://javaee.github.io/javaee-spec/javadocs/javax/validation/constraints/NotEmpty.html)

**Review by Max-Pochepets — APPROVED** [review]

File: src/main/java/com/dev/cinema/controller/UserController.java [hunk 3]
```diff
@@ -22,7 +24,7 @@ public UserController(UserService userService, UserMapper userMapper) {
     }
 
     @GetMapping("/by-email")
-    public UserResponseDto get(@RequestParam String email) {
+    public UserResponseDto get(@RequestParam @Size(min = 11) @Valid String email) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/controller/UserController.java]: does it work without the steps listed in this article?
https://www.baeldung.com/spring-validate-requestparam-pathvariable

File: src/main/java/com/dev/cinema/model/dto/request/CinemaHallRequestDto.java (hunk 2 above)
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/model/dto/request/CinemaHallRequestDto.java]: is int supported by this annotation?

File: src/main/java/com/dev/cinema/model/dto/request/MovieSessionRequestDto.java [hunk 4]
```diff
@@ -1,8 +1,16 @@
 package com.dev.cinema.model.dto.request;
 
+import javax.validation.constraints.Min;
+import javax.validation.constraints.NotNull;
+
 public class MovieSessionRequestDto {
+    @NotNull
+    @Min(1)
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/model/dto/request/MovieSessionRequestDto.java]: same - this annotation does not support wrappers

File: src/main/java/com/dev/cinema/model/dto/request/MovieSessionRequestDto.java [hunk 5]
```diff
@@ -1,8 +1,16 @@
...
     private Long movieId;
+    @NotNull
+    @Min(1)
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/model/dto/request/MovieSessionRequestDto.java]: same

File: src/main/java/com/dev/cinema/lib/PasswordValidatorAnnotation.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+    Class<?>[] groups() default {};
+
+    Class<? extends Payload>[] payload() default {};
+
+    @Target({ ElementType.TYPE })
+    @Retention(RetentionPolicy.RUNTIME)
+    @interface List {
+        PasswordValidatorAnnotation[] value();
+    }
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/lib/PasswordValidatorAnnotation.java]: do you need this?

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**